lib/dangermattic/plugins/labels_checker.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+module Danger
+  # Plugin to check for labels in a PR.
+  class LabelsChecker < Plugin
+    DEFAULT_DO_NOT_MERGE_LABELS = [
+      'Do Not Merge'
+    ].freeze
+
+    # Checks if a PR is missing labels or is marked with labels for not merging.
+    # If labels are missing, the plugin will emit a warning. If a label indicating that the PR should not be merged is present,
+    # an error will be emitted, preventing the final PR merge.
+    #
+    # @param do_not_merge_labels [String] The possible labels indicating that a merge should not be allowed.
+    #   Defaults to DEFAULT_DO_NOT_MERGE_LABELS if not provided.
+    # @param required_labels [Array<String>] The list of RegExes that will prevent the merge if any matching label is present in the PR.
+    #   Defaults to an empty array if not provided.
+    # @param required_labels_warning [String] The warning message displayed if the required labels are not present.
+    #   Defaults to showing the provided label regexes.
+    #
+    # @return [void]
+    def check(do_not_merge_labels: DEFAULT_DO_NOT_MERGE_LABELS, required_labels: [], required_labels_warning: nil)
+      github_labels = danger.github.pr_labels
+
+      # A PR shouldn't be merged with the 'DO NOT MERGE' label
+      found_labels = github_labels.select do |github_label|
+        do_not_merge_labels.any? { |label| github_label.casecmp?(label) }
+      end
+
+      failure("This PR is tagged with #{csv_markdown_list(found_labels)} label(s).") unless found_labels.empty?
+
+      # warn if a PR is missing any of the required labels
+      missing_required_labels = required_labels.reject do |required_label|
+        github_labels.any? { |pr_label| pr_label =~ required_label }
+      end
+
+      return if missing_required_labels.empty?
+
+      missing_labels_str_list = missing_required_labels.map(&:source)
+      warn(required_labels_warning || "PR is missing label(s) matching: #{csv_markdown_list(missing_labels_str_list)}")
```

Comment:
Since the API calls them `required_labels` I'd expect those to actually be **_required_**, i.e. to make the PR fail with a `failure`, not a `warn`.

I can imagine some repos only wanting this as a warning though to be less strict. So maybe that should be a parameter in the API, like `fail_on_missing_labels: true`? Or alternatively, provide both `required_labels: []` and `recommended_labels: []` parameters, with `required_labels` triggering `failure` if they're missing while `recommended_labels` only triggering `warn` (that way the callers can mix both cases if they need to, making the plugin all more flexible).

wdyt?

Reply:
That's an interesting point. I do wonder how often it happens that we want a label to cause a failure (and potentially even preventing a merge) -- it feels too strict of a process, but of course in the end this could be up to the host projects to decide.

As I don't think this is happening now in our repos, perhaps this might be a case where we need to see how the usage is going to be and, if it makes sense (or if it in fact improves a process), we can add it?
It's not even like it's a lot more work, of course, but I think it would be good to keep the API simple to start.

Reply:
Here's [a case where missing a required label makes the check fail, not warn](https://github.com/Automattic/peril-settings/blob/master/org/issue/label.ts#L13).

_Granted, this check is only applied on issues, not PRs, but that's the same idea. In this case, I _think_ this was enforced because the QE team uses (used?) GitHub labels to classify the bugs reported in the app, and create stats and priorities based on those, hence them being strongly required here I guess._

Reply:
I see a report [here](https://github.com/wordpress-mobile/WordPress-Android/issues/18767#issuecomment-1633490431)! Though not sure what will be the implication of fail compared to a warning for issues (guessing no difference?

I'm gonna give the `recommended_labels` approach a try, it shouldn't complicate the API that much.
But that also makes me think that, if we also have the option to fail on certain label patterns, wouldn't that make the `do_not_merge_labels` parameter obsolete?

Reply:
Well the `do_not_merge_labels` work in the opposite way: if the label is set, then we `fail`. As opposed to `required_labels` which fail only if the label is _not_ set. So no, I don't think that wouldn't make it obsolete.

Reply:
Ah, of course, I got confused for a moment 🙃 Both the label regex arrays parameters would be checking for labels we WANT the PR to have, otherwise failing/warning; while the "do not merge" labels are for potentially preventing a PR to be merged, always failing if they are present.

Reply:
Committed some changes [here](https://github.com/Automattic/dangermattic/pull/7/commits/b1e600005b556ce481c4fdd7bed4d0415349d281) adding the new parameters.